File: ffi/rust/src/ctx_ins.rs
use crate::{
  ctx_in::CTxIn,
  ffi::{
    are_ctx_ins_equal,
    get_ctx_in_at,
    get_ctx_ins_size,
  },
  macros::{
    impl_value_raw_const_obj,
  },
};
use std::{
  ffi::c_void,
  fmt,
};

#[derive(Debug, PartialEq, Eq)]
pub enum Error {
  IndexOutOfRange { index: usize, max_index: usize },
}

impl std::error::Error for Error {}

impl fmt::Display for Error {
  fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
    match self {
      Error::IndexOutOfRange { index, max_index } =>
        write!(f, "Index {index} is out of range. Max index is {max_index}"),
    }
  }
}

/* obj is an opaque pointer to:
 
   struct BlsctCTxIns {
     std::vector<CTxIn>* vec;
   };
*/
#[derive(Debug, Eq)]
pub struct CTxIns {
  obj: *const c_void,
}

impl CTxIns {
  pub fn get_ctx_in_at(&self, i: usize) -> Result<CTxIn, Error> {
    if i >= self.len() {
      return Err(Error::IndexOutOfRange {
        index: i,
        max_index: self.len() - 1,
      });
    }
    let obj = unsafe { get_ctx_in_at(self.value(), i) };
    Ok(obj.into())
  }

  pub fn len(&self) -> usize {
    unsafe { get_ctx_ins_size(self.value()) }
  }

  impl_value_raw_const_obj!();
}

impl From<*const c_void> for CTxIns {
  fn from(obj: *const c_void) -> CTxIns {
    CTxIns { obj }
  }
}

impl PartialEq for CTxIns {
  fn eq(&self, other: &Self) -> bool {
    unsafe { are_ctx_ins_equal(self.value(), other.value()) }
  }
}

File: ffi/ts/src/ctxIns.ts
import {
  getCTxInAt,
  getCTxInsSize,
} from './blsct'

import { CTxIn } from './ctxIn'

export class CTxIns {
  private obj: any

  constructor(obj: any) {
    this.obj = obj
  }
  
  at = (i: number): CTxIn => {
    const size = this.size()
    if (i < 0 || i >= size) {
      throw new RangeError(`Index ${i} is out of bounds. the size is ${size}`)
    }
    const obj = getCTxInAt(this.obj, i)
    return new CTxIn(obj)
  }

  size = (): number => {
    return getCTxInsSize(this.obj)
  }
}

